wolfcrypt/src/chacha20_poly1305.c

```diff
@@ -293,4 +293,176 @@ int wc_ChaCha20Poly1305_Final(ChaChaPoly_Aead* aead,
     return ret;
 }
 
+#ifdef HAVE_XCHACHA
+
+int wc_XChaCha20Poly1305_Init(
+    ChaChaPoly_Aead *aead,
+    const byte *ad, word32 ad_len,
+    const byte *nonce, word32 nonce_len,
+    const byte *key, word32 key_len,
+    int isEncrypt)
+{
+    byte authKey[CHACHA20_POLY1305_AEAD_KEYSIZE];
+    int ret;
+
+    if ((ad == NULL) || (nonce == NULL) || (key == NULL))
+        return BAD_FUNC_ARG;
+
+    if ((key_len != CHACHA20_POLY1305_AEAD_KEYSIZE) ||
+        (nonce_len != XCHACHA20_POLY1305_AEAD_NONCE_SIZE))
+        return BAD_FUNC_ARG;
+
+    if ((ret = wc_XChacha_SetKey(&aead->chacha,
+                                 key, key_len,
+                                 nonce, nonce_len,
+                                 0 /* counter */)) < 0)
+        return ret;
+
+    XMEMSET(authKey, 0, sizeof authKey);
+
+    /* Create the Poly1305 key */
+    if ((ret = wc_Chacha_Process(&aead->chacha, authKey, authKey,
+                                 (word32)sizeof authKey)) < 0)
+        return ret;
+    /* advance to start of the next ChaCha block. */
+    wc_Chacha_purge_current_block(&aead->chacha);
+
+    /* Initialize Poly1305 context */
+    if ((ret = wc_Poly1305SetKey(&aead->poly, authKey,
+                                 (word32)sizeof authKey)) < 0)
+        return ret;
+
+    if ((ret = wc_Poly1305Update(&aead->poly, ad, (word32)ad_len)) < 0)
+        return ret;
+
+    if ((ret = wc_Poly1305_Pad(&aead->poly, (word32)ad_len)) < 0)
+        return ret;
+
+    aead->isEncrypt = (byte)isEncrypt;
+    aead->state = CHACHA20_POLY1305_STATE_AAD;
+
+    return 0;
+}
+
+static WC_INLINE int wc_XChaCha20Poly1305_crypt_oneshot(
```

Comment:
Does it really make sense to use inline here?

Reply:
Yup -- stack is at a premium in kernel where we'll be using these initially, and better yet, the optimizer will remove the unneeded code as a function of `isEncrypt`, since it's passed in as a numeric constant by the only two functions that call it.